File: frontend/src/api/mappers/Permission.mapper.ts
import { type IPermission, type PermissionType } from '@/models/Permission.model';
import { type IApiPermission } from '@/api/api-models/ApiPermission.type';

export const permissionMapper = {

    map(permission: IApiPermission): IPermission {
        return {
            type: permissionMapper.mapType(permission.type),
            name: permission.name,
        };
    },

    mapType(type: number): PermissionType {
        switch (type) {
            case 1:
                return 'Admin';
            case 2:
                return 'Creator';
            default:
                throw new Error(`Unable to map permission type: '${type}.`);
        }
    },

    mapTypeToApi(type: PermissionType): number {
        switch (type) {
            case 'Admin':
                return 1;
            case 'Creator':
                return 2;
            default:
                throw new Error(`Unable to map permission type: '${type}.`);
        }
    },

};

File: backend/ClickStitch/Models/Mappers/PermissionMapper.cs
using Data.Records;

namespace ClickStitch.Models.Mappers;

public static class PermissionMapper
{
    public static PermissionModel Map(PermissionRecord permission) => new()
    {
        Type = MapType(permission.Type),
        Name = permission.Name
    };

    private static ApiPermissionType MapType(PermissionType type) => type switch
    {
        PermissionType.Admin => ApiPermissionType.Admin,
        PermissionType.Creator => ApiPermissionType.Creator,
        PermissionType.Stitcher => ApiPermissionType.Stitcher,

        PermissionType.Unknown or _ => throw new NotSupportedException($"Unable to map permission type: '{type}'.")
    };
}